papers/libceed-paper/paper.rst

```diff
@@ -0,0 +1,72 @@
+stability properties offered by higher order discretizations. Nowadays, high-order numerical methods are widely used in Partial Differential Equation (PDE) solvers, but software packages that provide high-performance implementations have often been special-purpose and intrusive. In contrast, libCEED is light-weight, minimally intrusive, and very versatile. When high-order finite/spectral element discretizations are used, the resulting sparse matrix representation of a global operator is computationally expensive, with respect to both the memory transfer and floating point operations needed for its evaluation. libCEED's Application Programming Interface (API) provides the local action of an operator (linear or nonlinear) without assembling its sparse representation. The purely algebraic nature of libCEED allows efficient operator evaluations (selectable at runtime) and offers matrix-free preconditioning ingredients. While libCEED’s focus is on high-order finite elements, the approach with which it is designed is algebraic and thus applicable to other discretizations in factored form.
+
+.. figure:: TensorVsAssembly.png
+   :align: center
+   :figclass: bht
+
+   Comparison of memory transfer and floating point operations per degree of freedom for different representations of a linear operator for a PDE in 3D with :math:`b` components and variable coefficients arising due to Newton linearization of a material nonlinearity. The representation labeled as *tensor* computes metric terms on the fly and stores a compact representation of the linearization at quadrature points. The representation labeled as *tensor-qstore* pulls the metric terms into the stored representation. The *assembled* representation uses a (block) CSR format.
+
+In this work, we first introduce libCEED’s conceptual framework and C interface, and then illustrate its new Python interface, developed using the C Foreign Function Interface (CFFI) for Python. CFFI allows to reuse most of the C declarations and requires only a minimal adaptation of some of them. The C and Python APIs are mapped in a nearly 1:1 correspondence. For instance, data stored in the CeedVector structure are associated to arrays defined via the numpy package. In fact, since libCEED heavily relies on pointers and arrays to handle the data, a Python structure that resembles the C arrays is needed. In details, numpy arrays allow this correspondence obtained by passing the numpy array memory address as pointers to the libCEED C API.
```

Comment:
I'm not sure how much effort it would be, but we could try using numpy or gnumpy so we have clearer CUDA + Python support? If its easy to add, it would be worth mentioning. I can look into it.

Reply:
I think JAX is probably the most interesting approach for device arrays, and it would bring with it the best in Python automatic differentiation.

Reply:
If we can run some tests to verify, then we already have libCEED + Python + CUDA and can say so here, as you can use jax.numpy in place of numpy, if I'm understanding correctly.

Reply:
FFI aspects of `jax.numpy` are surely different (at least for device arrays).

BTW, JAX is sort of a combined successor to numba and autograd (in the sense that developers moved from both of those projects to JAX); it does both JIT and AD, both with device support. There might be very little work needed to run on `jax.numpy` arrays, but someone would need to experiment.